src-tauri/src/cli/prompts.rs

```diff
@@ -180,48 +181,96 @@ pub fn check_and_install_python(
     Ok(())
 }
 
-pub fn select_mirrors(mut config: Settings) -> Result<Settings, String> {
-    if (config.wizard_all_questions.unwrap_or_default()
-        || config.idf_mirror.is_none()
-        || config.is_default("idf_mirror"))
-        && config.non_interactive == Some(false)
-    {
-        config.idf_mirror = Some(generic_select(
-            "wizard.idf.mirror",
-            &idf_im_lib::get_idf_mirrors_list()
-                .iter()
-                .map(|&s| s.to_string())
-                .collect(),
-        )?)
+async fn select_single_mirror<FGet, FSet>(
+    config: &mut Settings,
+    field_name: &str,    // e.g. "idf_mirror"
+    get_value: FGet,     // e.g. |c: &Settings| &c.idf_mirror
+    set_value: FSet,     // e.g. |c: &mut Settings, v| c.idf_mirror = Some(v)
+    candidates: &[&str], // list of mirror URLs
+    wizard_key: &str,    // e.g. "wizard.idf.mirror"
+    log_prefix: &str,    // e.g. "IDF", "Tools", "PyPI"
+) -> Result<(), String>
+where
+    FGet: Fn(&Settings) -> &Option<String>,
+    FSet: Fn(&mut Settings, String),
+{
+    // Interactive by default when non_interactive is None
+    let interactive = !config.non_interactive.unwrap_or_default();
+    let wizard_all = config.wizard_all_questions.unwrap_or_default();
+    let current = get_value(config);
+    let needs_value = current.is_none() || config.is_default(field_name);
+
+    // Only measure mirror latency if we actually need a value (or wizard wants to ask)
+    if interactive && (wizard_all || needs_value) {
+        let entries = calculate_mirrors_latency(candidates).await;
+        let display = entries.iter().map(|e| {
+            if e.latency.is_none() {
+                format!("{} (timeout)", e.url)
+            } else {
+                format!("{} ({:?} ms)", e.url, e.latency.unwrap())
+                }
+            })
+            .collect::<Vec<String>>();
+        let selected = generic_select(wizard_key, &display)?;
+        let url = selected.split(" (").next().unwrap_or(&selected).to_string();
+        set_value(config, url);
+    } else if needs_value {
+        let entries = calculate_mirrors_latency(candidates).await;
+        if let Some(entry) = entries.first() {
+            if entry.latency.is_none() {
+                info!("Selected {log_prefix} mirror: {} (timeout)", entry.url);
```

Comment:
You just selected for the user mirror which timeouted and threat it as normal situation, with logging info ?